song/endpoint.go

```diff
@@ -0,0 +1,92 @@
+/*
+ *     Copyright (C) 2018  Ontario Institute for Cancer Research
+ *
+ *     This program is free software: you can redistribute it and/or modify
+ *     it under the terms of the GNU Affero General Public License as published by
+ *     the Free Software Foundation, either version 3 of the License, or
+ *     (at your option) any later version.
+ *
+ *     This program is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU Affero General Public License for more details.
+ *
+ *     You should have received a copy of the GNU Affero General Public License
+ *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package song
+
+import (
+	"net/url"
+	"path"
+)
+
+type Endpoint struct {
+	BaseURL *url.URL
+}
+
+func (s *Endpoint) makeURL(args ...string) url.URL {
+	requestURL := *s.BaseURL
+	requestURL.Path = path.Join(s.BaseURL.Path, path.Join(args...))
+	return requestURL
+}
+
+// Upload uploads the file contents and returns the response
+func (s *Endpoint) Upload(studyID string, async bool) url.URL {
+	var url = s.makeURL("upload", studyID)
```

Comment:
Do we need error checking of some sort on the "public" function in this file?

Reply:
The functions are "public" to the rest of the program; but our end users shouldn't ever need to use them. 

Unless our end users start making up ids out of thin air that happen to include html special characters in them, we don't really have any errors to check... I suppose we could, though...